File: src/main/webapp/tutorials/code/python/functions-scope.py
# Variable Scope
# Scope refers to the region where a variable is accessible.

# 1. Local Scope
def my_func():
    x = 300 # Local variable
    print(f"Inside function: {x}")

my_func()
# print(x) # This would raise an error because x is not defined globally

# 2. Global Scope
x = 300 # Global variable

def my_func_2():
    print(f"Inside function (global): {x}")

my_func_2()
print(f"Outside function: {x}")

# 3. Naming Conflict
x = 300

def my_func_3():
    x = 200 # Creates a new local variable x
    print(f"Inside function (local override): {x}")

my_func_3()
print(f"Outside function (global remains): {x}")

# 4. Global Keyword
def my_func_4():
    global x
    x = 200 # Modifies the global variable

my_func_4()
print(f"After global modification: {x}")

File: src/main/webapp/tutorials/code/php/functions-scope.php
<?php
// Function Scope and Static Variables

// Global scope
$globalVar = "I'm global";

function testScope()
{
    // Cannot access $globalVar directly
    // echo $globalVar;  // Error!

    // Must use global keyword
    global $globalVar;
    echo $globalVar;
    echo "\n";
}

testScope();

// Static variables
function counter()
{
    static $count = 0;
    $count++;
    return $count;
}

echo counter();  // 1
echo "\n";
echo counter();  // 2
echo "\n";
echo counter();  // 3
echo "\n";

// Local scope
function localExample()
{
    $local = "I'm local";
    return $local;
}

echo localExample();
// echo $local;  // Error! Not accessible outside function
